internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,203 @@
+package spectest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/reddit/baseplate.go/errorsbp"
+)
+
+const (
+	client = "client"
+	server = "server"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errClientServer   = fmt.Errorf("missing %q or %q as the second part of the metric name", client, server)
+	errLength         = errors.New("metric name should have a minimum of three parts")
+	errNotFound       = errors.New("metric not found")
+	errLabelNotFound  = errors.New("label not found")
+	errPrometheusLint = errors.New("prometheus GatherAndLint")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed from baseplate.go
+// conform to the baseplate spec.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	t.Helper()
+
+	var batch errorsbp.Batch
+	batch.AddPrefix("validate spec", validateSpec(metricPrefix))
+
+	if len(batch.GetErrors()) > 0 {
+		t.Errorf(batch.Compile().Error())
+	}
+}
+
+func validateSpec(prefix string) errorsbp.Batch {
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return batch
+	}
+
+	allMetricNames := buildMetricNames(prefix)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
+		if _, ok := allMetricNames[name]; !ok {
+			continue
+		}
+
+		batch.Add(validatePromLint(name))
+		batch.Add(validateName(name, prefix))
+
+		labels := map[string]string{}
+		for _, metric := range m.GetMetric() {
+			for _, label := range metric.Label {
+				labels[*label.Name] = *label.Value
+			}
+		}
+		batch.Add(validateLabels(name, labels))
+
+		// after we validate the metric, delete it from the set of all
+		// expected metrics so that we can track if any were not found.
+		delete(allMetricNames, name)
+	}
+
+	if len(allMetricNames) > 0 {
+		notFound := make([]string, 0, len(allMetricNames))
+		for k := range allMetricNames {
+			notFound = append(notFound, k)
+		}
+		batch.Add(fmt.Errorf("%w metrics %v", errNotFound, notFound))
+	}
+
+	return batch
+}
+
+// buildMetricNames creates a set of all the expected metrics names for the given prefix
+// that should be registered with Prometheus as defined in the baseplate spec.
+// The prefix will be either thrift, http, or grpc.
+func buildMetricNames(prefix string) map[string]struct{} {
+	suffixes := []string{"_latency_seconds", "_requests_total", "_active_requests"}
+	names := map[string]struct{}{}
+	for _, suffix := range suffixes {
+		names[prefix+"_"+server+suffix] = struct{}{}
```

Comment:
🔕 
I think I would do `prefix + "_" + server + "_" + suffix` (i.e. remove the _ from the suffixes.

If you _really_ want to get fancy you can call

 https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus#BuildFQName

`names[prometheus.BuildFQName(prefix, client, suffix)]`